lib/pure/options.nim

```diff
@@ -191,6 +290,145 @@ proc `$`*[T](self: Option[T]): string =
   else:
     "None[" & name(T) & "]"
 
+macro optionCase*[T](m: Option[T], body: untyped): untyped =
+  ## A macro which provides a safe access pattern to
+  ## the Option type. In branches where the Option is not ``some`` then the
+  ## value is not available.
+  ##
+  ## ``optionCase`` makes the following conversion
+  ##
+  ## optionCase returnsOption():
+  ##    some x:
+  ##      <expr1 using x>
+  ##    none:
+  ##      <expr2 that does not use x>
+  ##
+  ## converts to --->
+  ## let m = returnsOption()
+  ## if m.isSome:
+  ##   let x = m.unsafeGet()
+  ##   <expr1 using x>
+  ## else:
+  ##   <expr2>
+  ##
+  assert body.len == 2
+  let
+    justHead = body[0][0]
+    nothingHead = body[1][0]
+
+  assert $justHead == "some",
+    "First block must be `some`: keyword not found"
+  assert $nothingHead == "none",
+    "Second block must be `none`: keyword not found"
+
+  let
+    mVal = genSym(nskLet)
+    validExpr = newDotExpr(mVal, ident("isSome"))
+    valueExpr = newDotExpr(mVal, ident("unsafeGet"))
+    justClause = nnkStmtList.newTree(
+      nnkLetSection.newTree(
+        nnkIdentDefs.newTree(
+          body[0][1],
+          newEmptyNode(),
+          valueExpr
+        )
+      ),
+      body[0][2]
+    )
+    nothingClause = body[1][1]
+
+  var ifExpr = newNimNode(nnkIfExpr)
+  ifExpr.add(newNimNode(nnkElifExpr).add(validExpr, justClause))
+  ifExpr.add(newNimNode(nnkElseExpr).add(nothingClause))
+
+  result = nnkStmtList.newTree(
+    nnkLetSection.newTree(
+      nnkIdentDefs.newTree(
+        mVal,
+        newEmptyNode(),
+        m
+      )
+    ),
+    ifExpr
+  )
+
+proc `>>=`*[T,U](self: Option[U], p: proc(value: U): Option[T]): Option[T] =
+  ## Used for chaining monadic computations together. Will create a none if
```

Comment:
Could you add a runnable example for these procedures?

Reply:
I personally don't like the `>>=` operator. I know it is used in haskell, but I think it is very ugly in Nim.

Reply:
I've completely rewritten most of it now, and it's a lot clearer. Tried to add runnableExamples, but it messed up when using macros so I converted them to code-blocks for now.